### ISSUE TITLE
Graph - Edges lost orientation after added to object

### ISSUE THREAD

**Issue by SMorrison42** [networkx/networkx#6015]
Hello,

I’ve create a Graph() object and add in the edge path I would like for it to follow. However with I print the object I notice that the edge orientation is swapped for a few of the entries in the graph. This in return impacts the calculation I perform to find the longest path. How do I preserve the order the edges are added to the graph? This is important for me to recover the correct path.

Example:

Here add to graph 5 4
Here add to graph **3 5**
Here add to graph 4 2
Here add to graph 4 1
Here add to graph **6 2**
Here add to graph **6 1**
Tgraph [('5', '4'), **('5', '3')**, ('4', '2'), ('4', '1'), **('2', '6'**), **('1', '6')**]

Notice that the ones in bold do not maintain edge orientation order

Then I use:
T = nx.dfs_tree(graph,1) = to start the traversal with the 1st element named ‘1’
longest_path = nx.dag_longest_path(T)
shortest_path = nx.dijkstra_path(graph,longest_path[0],longest_path[-1])

This is what is returned:
Longest_path = ['1', '4', '5', '3']
Shortest_path = ['1', '4', '5', '3']
Which is incorrect.

I believe the issue begins with the incorrect edge swap. Can someone assist me in how to maintain the order?

Best,
Shatavia

**Comment by dschult** [comment]
Perhaps you want a DiGraph instead of a Graph.  Directed graphs make a distinction between (u, v) and (v, u). Undirected graphs treat these as the same edge.

**Comment by rossbar** [comment]
It's important to recognize that you're not operating on the original graph, but on the tree that results from calling `nx.dfs_tree(graph, "1")`. This is a different graph object entirely - it *is* a DiGraph (which is why `dag_longest_path` even works, as that would raise for undirected inputs). This is easiest to get a handle on with a quick visualization:

#### The original graph, `G`
![G](https://user-images.githubusercontent.com/1268991/194133239-136a10db-8ae1-4560-bc31-8beba502ed7c.png)

#### The result of `nx.dfs_tree(G, "1")` (let's call it `T`)

![T](https://user-images.githubusercontent.com/1268991/194133459-deba815b-6fd3-48be-afe4-fb68e59b9f55.png)

From the visual inspection of the second graph `T`, it's clear that there are two valid longest paths: `1->4->5->3` and `1->4->2->6`, both composed of 3 edges. Note that `dag_longest_path` only returns *one* instance of the longest path in the case when there are multiple valid longest paths. You can use the `topo_order` kwarg of `dag_longest_path` to have it give you a different longest path by passing in a different (valid) topological order. For example:

```python
>>> nx.dag_longest_path(T)  # Gives one of the valid longest paths
['1', '4', '5', '3']
>>> nx.dag_longest_path(T, topo_order=["1", "4", "2", "5", "6", "3"])  # Different topological ordering results in a different, equally valid longest path
['1', '4', '2', '6']
```

Note also that there are multiple valid outputs from `dfs_tree` as well! For example, `1->6->2->4->5->3` is *also* a valid DFS tree. The output here *does* depend on the order that the edges are specified in the list:

```python
# Original specification
>>> G = nx.Graph([("5", "4"), ("5", "3"), ("4", "2"), ("4", "1"), ("2", "6"), ("1", "6")])
>>> nx.dfs_tree(G, "1").edges  # This is the image `T` above
OutEdgeView([('1', '4'), ('4', '5'), ('4', '2'), ('5', '3'), ('2', '6')])
# Same graph G, but move the last two edges to the front of the edge list
>>> G = nx.Graph([("1", "6"), ("2", "6"), ("5", "4"), ("5", "3"), ("4", "2"), ("4", "1")])
>>> nx.dfs_tree(G, "1").edges  # Different, but still valid, dfs tree
OutEdgeView([('1', '6'), ('6', '2'), ('2', '4'), ('4', '5'), ('5', '3')])
```

In summary, I think there are three main factors that contribute to the confusion in the original post:
 1. The path-finding is being done on the DFS tree, not the original graph
 2. There are multiple valid DFS trees - maybe you were getting a different one than you were expecting
 3. There can be multiple valid longest-paths in the DFS tree, and maybe you were getting a different one than you were expecting.

**Comment by SMorrison42** [comment]
Thank you so much for the explanation. This is extremely helpful and I’ll share with my colleagues. 

For additional information about our use with the package is that we are trying to use the package to help us improve on genome assemblies for bacterial and viral species. One of the issues we are running into is that one of the species has repetitive elements on the ends that are identical. An example is 3,4,5,1,2,5,4,3 with contig 1 typical being the longest contig we use as an anchor point and the edges are what we use to order the contig sequences. Our expected result for the example previously provided would of been 3,5,4,1,6,2,4,5,3. We will continue to explore the package and it’s functionality. Again we appreciate the explanation. 

Best,
Shatavia

**Comment by MridulS** [comment]
I'll go ahead and close this. thanks for opening this @SMorrison42
